server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -97,6 +155,131 @@ impl<Q: QueryExecutor + 'static> RemoteEngineServiceImpl<Q> {
         Ok(ReceiverStream::new(rx))
     }
 
+    async fn deduped_stream_read_internal(
+        &self,
+        request_notifiers: Arc<RequestNotifiers<StreamReadReqKey, Result<RecordBatch>>>,
+        request: Request<ReadRequest>,
+    ) -> Result<ReceiverStream<Result<RecordBatch>>> {
+        let instant = Instant::now();
+        let ctx = self.handler_ctx();
+        let (tx, rx) = mpsc::channel(STREAM_QUERY_CHANNEL_LEN);
+
+        let request = request.into_inner();
+        let table_engine::remote::model::ReadRequest {
+            table,
+            read_request,
+        } = request.clone().try_into().box_err().context(ErrWithCause {
+            code: StatusCode::BadRequest,
+            msg: "fail to convert read request",
+        })?;
+
+        let request_key = StreamReadReqKey::new(
+            table.table,
+            read_request.predicate.clone(),
+            read_request.projected_schema.projection(),
+            read_request.order,
+        );
+
+        let mut guard = match request_notifiers.insert_notifier(request_key.clone(), tx) {
+            // The first request, need to handle it, and then notify the other requests.
+            RequestResult::First => {
+                // This is used to remove key when future is cancelled.
+                ExecutionGuard::new(|| {
+                    request_notifiers.take_notifiers(&request_key);
+                })
+            }
+            // The request is waiting for the result of first request.
+            RequestResult::Wait => {
+                REMOTE_ENGINE_GRPC_HANDLER_DURATION_HISTOGRAM_VEC
```

Comment:
This observation is wrong because the query is not finished at this point. A better way to collect the metrics when `rx` is dropped.